File: native/android/app/src/main/kotlin/art/lazying/auspice/screens/Tarot.kt
package art.lazying.auspice.screens

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import art.lazying.auspice.*
import kotlinx.coroutines.delay
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

@Composable
fun TarotScreen(navController: NavController) {
    var spreadId by remember { mutableStateOf("three") }
    var question by remember { mutableStateOf("") }
    var draw by remember { mutableStateOf<TarotDraw?>(null) }
    var revealed by remember { mutableStateOf(setOf<String>()) }
    var error by remember { mutableStateOf<String?>(null) }
    var dealCount by remember { mutableIntStateOf(0) }

    LaunchedEffect(dealCount) {
        if (dealCount == 0) return@LaunchedEffect
        runCatching {
            Engines.evaluateAs<TarotDraw>("tarot.draw", buildJsonObject {
                put("spread", JsonPrimitive(spreadId))
                if (question.isNotBlank()) put("question", JsonPrimitive(question))
            })
        }.onSuccess {
            draw = it
            revealed = emptySet()
            error = null
            delay(550)
            it.cards.firstOrNull()?.let { first -> revealed = revealed + first.position.id }
        }.onFailure { error = l("This reading could not be computed. Please try again.") }
    }

    ScreenScaffold(
        navController,
        eyebrow = t("practice.tarot"),
        title = t("tarot.title"),
        tagline = t("tarot.tagline")
    ) {
        Panel {
            FlowRowOf {
                Chip(l("One card"), null, spreadId == "one") { spreadId = "one"; draw = null }
                Chip(l("Past · Present · Future"), null, spreadId == "three") { spreadId = "three"; draw = null }
                Chip(l("Celtic cross"), null, spreadId == "celtic") { spreadId = "celtic"; draw = null }
            }
            FieldLabel(t("tarot.question"))
            AuspiceField(question) { question = it }
            PrimaryButton(if (draw == null) t("tarot.draw") else t("tarot.drawAgain")) { dealCount++ }
        }

        error?.let { Panel(title = t("common.notComputed")) { Text(it, style = Type.serif(16), color = Palette.inkSoft) } }

        draw?.let { result ->
            ExplainReading(Json.encodeToString(result))
            Panel {
                SpreadStage(result, revealed) { card ->
                    revealed = revealed + card.position.id
                }
                if (revealed.size < result.cards.size) {
                    Text(
                        t("tarot.tapToTurn"),
                        style = Type.sans(13), color = Palette.inkMute,
                        modifier = Modifier.fillMaxWidth(), textAlign = TextAlign.Center
                    )
                }
            }
            result.cards.filter { revealed.contains(it.position.id) }.forEach { CardReading(it) }
        }
    }
}

@Composable
fun AuspiceField(value: String, onChange: (String) -> Unit) {
    Box(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(Color(0x40000000))
            .border(1.dp, Palette.line, RoundedCornerShape(14.dp))
            .padding(horizontal = 14.dp, vertical = 12.dp)
    ) {
        BasicTextField(
            value = value,
            onValueChange = onChange,
            textStyle = Type.serif(18).copy(color = Palette.ink),
            cursorBrush = SolidColor(Palette.gold),
            modifier = Modifier.fillMaxWidth()
        )
    }
}

/** The cards, placed by the spread's own 0–1 layout grid. */
@Composable
private fun SpreadStage(draw: TarotDraw, revealed: Set<String>, onTap: (DrawnCard) -> Unit) {
    val tall = draw.cards.size > 3
    val cardWidth = if (tall) 56.dp else 88.dp
    Layout(
        content = {
            draw.cards.forEach { card ->
                Box(Modifier.graphicsLayer {
                    rotationZ = if (card.position.layout.rotate == true) 90f else 0f
                }) {
                    TarotCardView(card, revealed.contains(card.position.id), cardWidth) { onTap(card) }
                }
            }
        },
        modifier = Modifier
            .fillMaxWidth()
            .height(if (tall) 420.dp else if (draw.cards.size == 1) 220.dp else 200.dp)
    ) { measurables, constraints ->
        val placeables = measurables.map { it.measure(constraints.copy(minWidth = 0, minHeight = 0)) }
        layout(constraints.maxWidth, constraints.maxHeight) {
            placeables.forEachIndexed { index, placeable ->
                val layout = draw.cards[index].position.layout
                placeable.place(
                    x = (layout.x * constraints.maxWidth - placeable.width / 2).toInt(),
                    y = (layout.y * constraints.maxHeight - placeable.height / 2).toInt()
                )
            }
        }
    }
}

/** One card, with a real turn rather than a cross-fade. */
@Composable
fun TarotCardView(card: DrawnCard, faceUp: Boolean, width: androidx.compose.ui.unit.Dp, onTap: () -> Unit) {
    val turn by animateFloatAsState(if (faceUp) 180f else 0f, tween(550), label = "turn")
    Box(
        Modifier
            .size(width, width * 1.62f)
            .graphicsLayer {
                rotationY = turn
                cameraDistance = 14f * density
            }
            .clip(RoundedCornerShape(9.dp))
            .clickable(onClick = onTap)
    ) {
        if (turn < 90f) {
            Box(
                Modifier
                    .fillMaxSize()
                    .background(Brush.linearGradient(listOf(Palette.night3, Palette.night2)))
                    .border(1.dp, Palette.goldLine, RoundedCornerShape(9.dp)),
                contentAlignment = Alignment.Center
            ) {
                Text("✦", style = Type.display((width.value * 0.3f).toInt()), color = Palette.gold.copy(alpha = 0.7f))
            }
        } else {
            Box(
                Modifier
                    .fillMaxSize()
                    .graphicsLayer { rotationY = 180f }
                    .background(Brush.verticalGradient(listOf(Palette.parchment, Palette.parchment2)))
                    .border(1.dp, Color(0x808A6A22), RoundedCornerShape(9.dp))
                    .graphicsLayer { rotationZ = if (card.reversed) 180f else 0f }
                    .padding(horizontal = 5.dp, vertical = 8.dp),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(3.dp)
                ) {
                    Text(card.card.label, style = Type.display((width.value * 0.14f).toInt(), FontWeight.Bold), color = Color(0xFF6B4E16))
                    Text(elementGlyph(card.card.element), style = Type.display((width.value * 0.24f).toInt()), color = Color(0xFF8A6A22))
                    Text(l(card.card.text.en.name),
                        style = Type.display((width.value * 0.115f).toInt()),
                        color = Color(0xFF3A2A08),
                        textAlign = TextAlign.Center,
                        maxLines = 2
                    )
                }
            }
        }
    }
}

private fun elementGlyph(element: String) = when (element) {
    "fire" -> "🜂"
    "water" -> "🜄"
    "air" -> "🜁"
    "earth" -> "🜃"
    else -> "✶"
}

@Composable
private fun CardReading(card: DrawnCard) {
    Panel {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(l(card.position.name.en),
                style = Type.display(15).copy(letterSpacing = 1.6.sp),
                color = Palette.gold
            )
            Spacer(Modifier.width(8.dp))
            Spacer(Modifier.weight(1f))
            if (card.reversed) {
                Text(
                    t("tarot.reversed"),
                    style = Type.sans(10, FontWeight.Bold).copy(letterSpacing = 1.6.sp),
                    color = Palette.rose
                )
            }
        }
        Text(l(card.card.text.en.name), style = Type.display(21), color = Palette.ink)
        Text(l(card.position.question.en),
            style = Type.serif(16).copy(fontStyle = FontStyle.Italic),
            color = Palette.inkMute
        )
        FlowRowOf(spacing = 6) {
            val words = if (card.reversed) {
                l(card.card.text.en.reversed.joinToString(" · ")).split(" · ")
            } else {
                l(card.card.text.en.upright.joinToString(" · ")).split(" · ")
            }
            words.forEach { word ->
                Text(l(word),
                    style = Type.sans(13, FontWeight.SemiBold),
                    color = Palette.inkSoft,
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(Color(0x0DFFFFFF))
                        .padding(horizontal = 10.dp, vertical = 6.dp)
                )
            }
        }
    }
}

File: native/ios/Auspice/Screens/Tarot.swift
import SwiftUI

struct TarotScreen: View {
    @State private var spreadId = "three"
    @State private var question = ""
    @FocusState private var questionFocused: Bool
    @State private var draw: TarotDraw?
    @State private var revealed: Set<String> = []
    @State private var dealing = false
    @State private var selected: DrawnCard?
    @State private var error: String?

    private let spreads: [(id: String, label: String, chinese: String, count: Int)] = [
        ("one", "One card", "单张", 1),
        ("three", "Past · Present · Future", "三张", 3),
        ("celtic", "Celtic cross", "凯尔特十字", 10)
    ]

    var body: some View {
        ScreenScaffold(
            eyebrow: t("practice.tarot"),
            title: t("tarot.title"),
            tagline: t("tarot.tagline")
        ) {
            Panel {
                FlowRow(spacing: 8) {
                    ForEach(spreads, id: \.id) { spread in
                        Chip(
                            label: l(spread.label),
                            active: spreadId == spread.id
                        ) {
                            spreadId = spread.id
                            draw = nil
                            revealed = []
                        }
                    }
                }
                FieldLabel(t("tarot.question"))
                TextField("", text: $question, axis: .vertical)
                    .accessibilityIdentifier("tarot.question")
                    .focused($questionFocused)
                    .textFieldStyle(AuspiceFieldStyle())
                    .lineLimit(1...3)
                Button(action: deal) {
                    Label(draw == nil ? t("tarot.draw") : t("tarot.drawAgain"), systemImage: "sparkles")
                }
                .buttonStyle(PrimaryButtonStyle())
                .accessibilityIdentifier("tarot.compute")
                .disabled(dealing)
            }

            if let error {
                Panel(title: t("common.notComputed")) {
                    Text(error).font(Typeface.serif(16)).foregroundStyle(Palette.inkSoft)
                }
            }

            if let draw {
                ExplainReading(result: draw).id(draw.seed)
                Panel {
                    SpreadStage(draw: draw, revealed: revealed, tap: { card in
                        if revealed.contains(card.id) {
                            selected = card
                        } else {
                            _ = withAnimation(.spring(response: 0.55, dampingFraction: 0.75)) {
                                revealed.insert(card.id)
                            }
                        }
                    })
                    .frame(height: stageHeight(draw.spread.id))
                    if revealed.count < draw.cards.count {
                        Text(t("tarot.tapToTurn"))
                            .font(Typeface.sans(13))
                            .foregroundStyle(Palette.inkMute)
                            .frame(maxWidth: .infinity)
                    }
                }

                ForEach(draw.cards.filter { revealed.contains($0.id) }) { card in
                    CardReading(card: card)
                }
            }
        }
        .sheet(item: $selected) { card in
            CardDetail(card: card)
        }
    }

    private func stageHeight(_ spread: String) -> CGFloat {
        switch spread {
        case "one": 230
        case "three": 210
        default: 430
        }
    }

    private func deal() {
        questionFocused = false
        dealing = true
        revealed = []
        do {
            var input: [String: Any] = ["spread": spreadId]
            if !question.trimmingCharacters(in: .whitespaces).isEmpty { input["question"] = question }
            let result = try Engines.shared.evaluate("tarot.draw", input, as: TarotDraw.self)
            error = nil
            withAnimation(.easeOut(duration: 0.3)) { draw = result }
            // Deal them face down, then turn the first after a beat, the way a
            // reader lays the spread out before saying anything.
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.55) {
                if let first = result.cards.first {
                    withAnimation(.spring(response: 0.6, dampingFraction: 0.75)) {
                        _ = revealed.insert(first.id)
                    }
                }
                dealing = false
            }
        } catch {
            self.error = l("This reading could not be computed. Please try again.")
            dealing = false
        }
    }
}

/// The cards, placed by the spread's own 0–1 layout grid.
private struct SpreadStage: View {
    let draw: TarotDraw
    let revealed: Set<String>
    let tap: (DrawnCard) -> Void

    var body: some View {
        GeometryReader { geometry in
            ZStack {
                ForEach(Array(draw.cards.enumerated()), id: \.element.id) { index, card in
                    TarotCardView(
                        card: card,
                        faceUp: revealed.contains(card.id),
                        size: cardSize(geometry.size)
                    )
                    .rotationEffect(.degrees(card.position.layout.rotate == true ? 90 : 0))
                    .position(
                        x: card.position.layout.x * geometry.size.width,
                        y: card.position.layout.y * geometry.size.height
                    )
                    .onTapGesture { tap(card) }
                    .transition(.opacity.combined(with: .scale(scale: 0.8)))
                    .animation(
                        .spring(response: 0.5, dampingFraction: 0.8).delay(Double(index) * 0.07),
                        value: draw.seed
                    )
                }
            }
        }
    }

    private func cardSize(_ available: CGSize) -> CGSize {
        let count = draw.cards.count
        let width: CGFloat = count > 3 ? 58 : min(94, available.width / CGFloat(count) - 14)
        return CGSize(width: width, height: width * 1.62)
    }
}

/// One card, with a real three-dimensional turn rather than a cross-fade.
struct TarotCardView: View {
    let card: DrawnCard
    let faceUp: Bool
    let size: CGSize

    var body: some View {
        ZStack {
            back.opacity(faceUp ? 0 : 1)
            face
                .opacity(faceUp ? 1 : 0)
                .rotation3DEffect(.degrees(180), axis: (x: 0, y: 1, z: 0))
        }
        .frame(width: size.width, height: size.height)
        .rotation3DEffect(.degrees(faceUp ? 180 : 0), axis: (x: 0, y: 1, z: 0), perspective: 0.45)
        .shadow(color: .black.opacity(0.45), radius: 10, y: 6)
    }

    private var back: some View {
        RoundedRectangle(cornerRadius: 9, style: .continuous)
            .fill(
                LinearGradient(
                    colors: [Palette.night3, Palette.night2],
                    startPoint: .topLeading,
                    endPoint: .bottomTrailing
                )
            )
            .overlay(
                RoundedRectangle(cornerRadius: 9, style: .continuous)
                    .strokeBorder(Palette.goldLine, lineWidth: 1)
            )
            .overlay(
                Image(systemName: "sparkle")
                    .font(.system(size: size.width * 0.3))
                    .foregroundStyle(Palette.gold.opacity(0.7))
            )
    }

    private var face: some View {
        RoundedRectangle(cornerRadius: 9, style: .continuous)
            .fill(
                LinearGradient(
                    colors: [Palette.parchment, Color(hex: 0xE9DBB9)],
                    startPoint: .top,
                    endPoint: .bottom
                )
            )
            .overlay(
                VStack(spacing: 3) {
                    Text(card.card.label)
                        .font(Typeface.display(size.width * 0.14, weight: .bold))
                        .foregroundStyle(Color(hex: 0x6B4E16))
                    Image(systemName: elementSymbol)
                        .font(.system(size: size.width * 0.26))
                        .foregroundStyle(Color(hex: 0x8A6A22))
                    Text(l(card.card.text.en.name))
                        .font(Typeface.display(size.width * 0.115))
                        .foregroundStyle(Color(hex: 0x3A2A08))
                        .multilineTextAlignment(.center)
                        .minimumScaleFactor(0.6)
                        .lineLimit(2)
                }
                .padding(.horizontal, 5)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 9, style: .continuous)
                    .strokeBorder(Color(hex: 0x8A6A22).opacity(0.5), lineWidth: 1)
            )
            .overlay(alignment: .bottomTrailing) {
                if card.reversed {
                    Image(systemName: "arrow.uturn.down")
                        .font(.system(size: size.width * 0.12, weight: .bold))
                        .foregroundStyle(Palette.rose)
                        .padding(4)
                }
            }
            .rotationEffect(.degrees(card.reversed ? 180 : 0))
    }

    private var elementSymbol: String {
        switch card.card.element {
        case "fire": "flame"
        case "water": "drop"
        case "air": "wind"
        case "earth": "leaf"
        default: "star"
        }
    }
}

/// What a turned card says in its position.
private struct CardReading: View {
    let card: DrawnCard

    var body: some View {
        Panel {
            HStack(alignment: .firstTextBaseline) {
                Text(l(card.position.name.en))
                    .font(Typeface.display(15))
                    .tracking(1.6)
                    .foregroundStyle(Palette.gold)
                Spacer()
                if card.reversed {
                    Text(t("tarot.reversed"))
                        .font(Typeface.sans(10, weight: .bold))
                        .tracking(1.6)
                        .foregroundStyle(Palette.rose)
                }
            }
            Text(l(card.card.text.en.name))
                .font(Typeface.display(21))
                .foregroundStyle(Palette.ink)
            Text(l(card.position.question.en))
                .font(Typeface.serif(16))
                .italic()
                .foregroundStyle(Palette.inkMute)
                .fixedSize(horizontal: false, vertical: true)
            FlowRow(spacing: 6) {
                ForEach(keywords, id: \.self) { word in
                    Text(l(word))
                        .font(Typeface.sans(13, weight: .semibold))
                        .foregroundStyle(Palette.inkSoft)
                        .padding(.horizontal, 10)
                        .padding(.vertical, 6)
                        .background(Capsule().fill(Color.white.opacity(0.05)))
                }
            }
        }
    }

    private var keywords: [String] {
        let english = card.reversed ? card.card.text.en.reversed : card.card.text.en.upright
        return l(english.joined(separator: " · ")).components(separatedBy: " · ")
    }
}

private struct CardDetail: View {
    let card: DrawnCard
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        ZStack {
            Sky()
            VStack(spacing: 18) {
                TarotCardView(card: card, faceUp: true, size: CGSize(width: 150, height: 243))
                CardReading(card: card)
                Spacer()
            }
            .padding(18)
        }
        .presentationBackground(Palette.night)
        .overlay(alignment: .topTrailing) {
            Button { dismiss() } label: {
                Image(systemName: "xmark.circle.fill")
                    .font(.system(size: 26))
                    .foregroundStyle(Palette.inkMute)
                    .padding(14)
            }
        }
    }
}
